### ISSUE TITLE
Snapshot performance predictability

### ISSUE THREAD

**Issue by AusIV** [ethereum/go-ethereum#21310]
#### System information

Geth version: `1.9.14-stable`
OS & Version: Linux

#### Expected behaviour

The performance benefits of the snapshot system introduced in 1.9.13 are pretty phenomenal - We've seen some eth_call queries go from 5+ seconds to well under a second with a fully loaded snapshot in place. Running an operation where capacity planning relies on predictable performance, we'd love to be able to take advantage of the snapshot, but to do so requires consistent, predictable performance.

#### Actual behaviour

In the event of an abrupt termination of Geth, the snapshot journal is not written to disk. On the next startup, because the journal is missing, the snapshot is wiped and begins regenerating. On the Ethereum mainnet, snapshot generation can take several days, during which performance is considerably below the peak performance operating with a completed snapshot.

#### Steps to reproduce the behaviour

kill -9 $GETH_PID

Restart Geth

#### Proposed Solution

In our case, it would be very preferable to have a longer startup time but have servers come online with working snapshots, rather than come online quickly but spend days reconstructing the snapshot. There are two possible ways to do this:

1. Our operation typically runs with `--gcmode=archive`. Given that we have all of the intermediate state tries between the 128 block old diskLayer and the current block, we should be able to reconstruct the 128 diffLayers by comparing the state tries while initializing the snapshot. We'd still use the journal for clean shutdowns, as it's much faster than reconstructing the diffLayer by comparing state tries.
2. For those who don't run with `--gcmode=archive` (which I believe is most users), it may be necessary to roll the current block back to the same block as the diskLayer, and resume syncing from there. I'm not sure how feasible this is, as it seems the snapshot tree calls Cap(root, 127) every block, while the trie only gets synced to disk when the dirty threshold is exceeded, and it seems unlikely that those events are likely to line up. If we could ensure that commits to the disk layer happen at the same time as the trie gets committed to disk, then unclean shutdowns should be able to resume from just the disk layer without having to discard the snapshot. Again, we'd still use the journal for clean shutdowns as that makes it easier to handle possible reorgs, but this approach gives us a reasonable hope of not having to discard the snapshot after an unclean shutdown.

I've already been working on a proof-of-concept of option 1 - once I have the kinks worked out I'd be happy to push it upstream if you'd be amenable. I'd also be willing to hash out a PR for option 2 if that's something you'd be open to, but since I don't need it myself I don't want to sink the time into it if the Geth team has other plans.

**Comment by holiman** [comment]
Nice to hear you're experimenting with snapshot! And also good to hear it's "pretty phenomenal", thanks!

I agree that there should be a way to start geth in "finish up the snapshot, _then_ to the blockchain stuff". 
Regarding your proposals: 

The problem is not so much that we need to reconstruct 128 layers, geth can handle it fine if those layers are missing (except that we won't be able to handle reorgs nicely). Actually, after a snapshot is newly generated, those memory layers are not even present -- there's only the disklayer. 

I want to clarify one thing re flushing to disk. So, there are only 127 memory-layers (or 128?), but those do not map 1:1 to 128 previous states/blocks. They map to _at minimum_ 128 blocks, but we merge into the bottom-most memory layer, so that layer can represent N (maybe 100) actual layers. So the guarantee is that we can reorg to any of the latest 128 states, but we accumulate older blocks into the bottom-most mem layer, and every once in a while flush that one to disk. So

There are a couple of things that could be done: 

1. Create snapshot before starting blockchain would be a good feature. 
2. Flush snapshots to disk every time we also flush the trie. However, this won't work well, since the next snapshot-flush will un-align these anyway. 
3. Flush trie to disk every time we flush the snapshot. This won't work well either, since it'll screw up the pruning, and eat even more disk space.

Anyway, I think your proposal number 1 won't really work, because it's not sufficient to create 128 block-states on top of the disklayer -- but if you go through _all_ block-states and merge the bottom ones into the bottom layer, then it could work.

**Comment by AusIV** [comment]
Sorry, when I said "reconstruct 128 layers" what I really meant is to reconstruct the layers between the diskLayer and the latest block - in my experimenting that has been 128 layers, but I'm also working with my fork of Geth that does streaming replication, and I don't think we're consolidating changes in the bottom-most memory layer - we're just merging the bottom layer to disk with each successive block (we'll need to revisit that separately). The pseudocode of what I've done is basically:

```
header = ReadHeadHeader()
root = GetSnapshotRoot(diskdb)
intermediateRoots = []common.Hash{header.Root}
for header.Root != root {
    header = ReadHeader(header.ParentHash)
    intermediateRoots = append([]common.Hash{header.Root}, intermediateRoots...)
}

parentRoot = root
for i = 1; i < len(intermediateRoots); i++ {
    destructs, accounts, storage, err = DiffTries(intermediateRoots[i-1], intermediateRoots[i])
    // handle error
    snaps.Update(intermediateRoots[i], intermediateRoots[i-1], destructs, accounts, storage)
}
```

So I'm not hard coding anything to 128 blocks - it's just that in my experience with my replication system that number has been 128. 

As far as your list of options: Number 1 is useful for initializing things, but in my experience generating an initial snapshot from scratch takes several days, so isn't an answer when you're spinning up instances to deal with capacity limitations. Would it not be possible to combine options 2 and 3? If we synchronize the snapshot flushes with the trie writes such that they all get written at once (never flushing the trie to disk without flushing the snapshot and never flushing the snapshot to disk without flushing the trie). Is it necessary to to flush the snapshots to disk more often than we're flushing the trie to disk? If we can keep those two things in sync we can avoid having to rebuild the snapshot or suffering a performance degradation as it rebuilds.

**Comment by karalabe** [comment]
As Martin mentioned, Option 1 will not work because the disk layer might be older than 128. It would get a bit messy and I don't want to rely on - or have special casing - for archive nodes.

Option 2 is much more interesting though. Even if the persistent disk layer is lets say 200-300 blocks old, we *could* rerun those blocks on startup. Although the state trie is already pruned, we *know* that the blocks we have in the database are all already validated, so we could just run via the snapshots and skip merkle root checks + trie commits for any blocks that we don't have the trie for. This would allow us to even faster crunch through the blocks and reconstruct the missing diff layers.

**Comment by karalabe** [comment]
Hmm, that said, I do see the issue with the trie/snap desync. If I murder Geth, I could recreate the snapshot diffs easily but the trie might be completely missing. Hmm.

We can't afford to keep that many diff layers in RAM. Tries are pruned on the fly and are pushed to disk rarely in their entirety (maybe once a day). That works ok-ish because if we run out of memory allowance, we can just flush overflow items which are old, still keeping data consistent. In the case of the snapshots, we can't flush just bits here and there, we'd need to push entire layers to keep the entire disk layer consistent.

**Comment by karalabe** [comment]
Perhaps a hard but doable solution would be to explicitly detect if the snapshot is available, but desynced, and in that case roll back the blocks to the most recent available state, reprocess all the blocks via trie only; then go back to the most recent available snapshot and reprocess snapshot only. It's annoying, but it would be infinitely faster than it is now (you need to reprocess the trie anyway, so no difference there, but the snapshot could be salvaged).

**Comment by AusIV** [comment]
That makes a lot of sense. To restate, the basic idea would be to identify that the snapshot is ahead of the state trie and set it aside until the block processing catches up to it, then once caught up resume updating the snapshot?

That would still require a special case for archive nodes, as in the case where Geth dies abruptly the trie will be ahead of the snapshot rather than behind it, though that doesn't necessarily have to be as complex a special case as I described earlier - it would just need to roll back the block to match the snapshot and resume from there rather try to bring the snapshot up to date.

Perhaps you could clarify one other thing for me - when a Geth node dies abruptly, upon restart it resets the head to the last available state, then reacquires subsequent blocks from peers, correct? It doesn't reprocess the blocks stored in the database, does it?

**Comment by holiman** [comment]
Coming back to this, sorry for the delay.

> Perhaps you could clarify one other thing for me - when a Geth node dies abruptly, upon restart it resets the head to the last available state, then reacquires subsequent blocks from peers, correct? It doesn't reprocess the blocks stored in the database, does it?

As of https://github.com/ethereum/go-ethereum/pull/21409, the behaviour upon restart with various types of corrupt data is a lot more well defined. I think your description matches the behaviour, basically:

- reset to last available state, 
- prune 'future' blocks if needed, 
- fetch blocks from network

@karalabe correct me if I'm wrong
